Add ELF library for manipulating ELF objects on the filesystem

joestringer opened the pull request:
Add a new package that is capable of reading ELF files and substituting
32-bit values in the data section and strings in the string table, based
on a caller-specified set of options.

This works by:
* `elf.Open(path)` - reads an ELF object from the filesystem and locates all relevant data/string offsets in the binary
* `elf.Write(path, dataOptions, stringOptions)` - takes a map of data options (symbol name to 32-bit value to substitute in the ELF), and a map of string options (substitute map key with map value in the ELF) and a path to write the new object to.

There's a new `elf-demo.c` BPF program under `test/bpf/` which demonstrates the basics of this functionality, and provides a basis for testing the implementation via golang unit tests. I have introduced a dependency in the `unit-tests` make target to ensure that `test/bpf` is compiled to make the ELF available during unit testing.

Currently this is not used, but #7095 will build on top of this package
to provide templating of BPF programs.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7196)
<!-- Reviewable:end -->

joestringer committed:
Add ELF package for templating constants

Add a new package that is capable of reading ELF files and substituting
32-bit values in the data section and strings in the string table, based
on a caller-specified set of options.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
bpf: Add fetch helpers to lib/utils.h

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer commented on pkg/elf/elf.go:
```diff
@@ -0,0 +1,260 @@
+// On success, writes the file to the specified path. On failure, returns an
+// error and no file is left on the filesystem.
+//
+// Concurrent writes from the same source ELF is NOT safe.
```
(from other PR):

nit: while concurrent write from the same ELF struct is not safe, concurrent copy from the same source object file on the filesystem should be fine. Workaround is to Open() the same object multiple times for concurrent access.

So long as this function returns successfully, one should be able to call `Write()` multiple times. Maybe we could put a lock in here and share the in-memory `*ELF` for each template. At the moment in the other PR we're basically just re-opening the template ELF from the endpoint regeneration code. This should be a simple incremental patch on top though, so unless someone feels strongly I'll skip this for now.

joestringer commented on pkg/elf/elf.go:
```diff
@@ -0,0 +1,260 @@
+}
+
+// Write the received ELF to a new file at the specified location, with the
+// specified options (indexed by name) substituted in the data section.
```
nit: And the specified string options substituted in the strtab.

joestringer replied:
Each map key and value in `strOptions` must be of equal length. I believe I left a warning log somewhere if this is not the case.

joestringer commented on pkg/elf/elf.go:
```diff
@@ -0,0 +1,260 @@
+}
+
+// IgnoreSymbolPrefixes configures the ELF package to ignore symbols that have
+// any of the specified prefixes. It must be called by only one thread at a time.
```
For context, this basically allows us to write a warning log whenever we attempt to write a new ELF without substituting one of the symbols. PR #6618 includes code to ignore all of the global maps symbols, because we don't need to substitute those.

aanm commented on pkg/elf/elf.go:
```diff
@@ -0,0 +1,260 @@
+}
+
+// newElf reads the ELF at the specified path.
+func newELF(f *os.File, path string) (*ELF, error) {
```
Wait a second, is it me or `path` is not even used to read the file?

joestringer replied:
Yeah this is a little ugly... let me think about how I can clean this up.

I had two intents here:
* Try to separate the filesystem access from the core ELF
* Ensure that errors that occur while handling the ELF have the context of the template path.

I think both of these can be achieved by pushing the filesystem handling out to the caller, and just pass a `io.ReaderAt` here.. then either pass all the errors out or pass the logger in.

aanm commented on pkg/elf/elf.go:
```diff
@@ -0,0 +1,260 @@
+func (elf *ELF) copy(w io.Writer, r io.ReadSeeker, intOptions map[string]uint32, strOptions map[string]string) error {
+	if len(intOptions) == 0 && len(strOptions) == 0 {
+		// Copy the remaining portion of the file
+		if _, err := io.Copy(w, r); err != nil {
```
`return io.Copy(w, r)`

joestringer replied:
`io.Copy()` returns multiple values, this function does not.

aanm commented on pkg/elf/elf.go:
```diff
@@ -0,0 +1,260 @@
+			v, exists := strOptions[symbol.name]
+			if exists {
+				if uint64(len(v)) != symbol.size {
+					return fmt.Errorf("symbol substitution value %q (len %d) must equal length of symbol name %q (len %d)", v, len(v), symbol.name, len(symbol.name))
```
the returned error message is giving the wrong information. you are comparing with `symbol.size` but the error message is stating `len(symbol.name)`

aanm commented on pkg/elf/elf.go:
```diff
@@ -0,0 +1,260 @@
+		if err := binary.Write(w, elf.ByteOrder, value); err != nil {
+			return fmt.Errorf("failed to substitute %s: %s", symbol.name, err)
+		}
+		if _, err := r.Seek(int64(symbol.size), io.SeekCurrent); err != nil {
```
Is this it safe for parallel access?

joestringer replied:
It wasn't, I did a bit of a rewrite and now I think it is (but the users aren't yet attempting to use it in parallel).

aanm commented on pkg/elf/elf.go:
```diff
@@ -0,0 +1,260 @@
+	}
+
+	// Copy the remaining portion of the file
+	if _, err := io.Copy(w, r); err != nil {
```
`return io.Copy(w, r)`

joestringer replied:
`io.Copy()` returns multiple values, including the number of bytes written, so we can't directly return here. Even if we change the return signature and `return io.Copy(...)`  here, we would be returning the wrong number of bytes.

aanm commented on pkg/elf/elf.go:
```diff
@@ -0,0 +1,260 @@
+	if err == nil {
+		err = f.Sync()
+	}
+	if _, err2 := elf.file.Seek(0, io.SeekStart); err2 != nil {
```
FYI, since you are assigning a new variable you can still use `err` and the value set in line 231 won't be lost

joestringer replied:
Nice to know, but this kind of accidental hiding is really easy to get wrong so I'd rather be a bit more explicit here.

joestringer replied:
I ended up removing this anyway to make it more (likely to be) safe for concurrent use.

aanm commented on pkg/elf/elf.go:
```diff
@@ -0,0 +1,260 @@
+		).Warning("Failed to seek to beginning of ELF")
+	}
+
+	f.Close()
```
f.Close() also returns an error

joestringer replied:
Ugh, yeah. I feel like there's no good way to properly handle this. I've tried another approach with a defer, let me know what you think.

aanm commented on pkg/elf/elf_test.go:
```diff
@@ -0,0 +1,236 @@
+	if err != nil {
+		return err
+	}
+
```
```golang
if !bytes.Equal(h1, h2) {
     return errDifferentFiles
}
```

aanm commented on pkg/elf/elf_test.go:
```diff
@@ -0,0 +1,236 @@
+		return "", fmt.Errorf("Failed to get working directory: %s", err)
+	}
+
+	return filepath.Join(wd, "..", "..", "test", "bpf", "elf-demo.o"), nil
```
can you guarantee it will always be this relative path? Isn't there a way that we could pass the elf-demo.o location via a global variable or something like it? :thinking:

joestringer replied:
Yes, I can guarantee it - the unit tests will fail otherwise ;)

In all seriousness though, I found a nice little Makefile/ldflags hack for this, I'll roll in an extra patch for this.

joestringer committed:
testutils: Store root directory path in unit testing

Store the root directory path in testutils for use throughout the tree.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
loader: Simplify unit tests with CiliumRootDir

Update the loader tests to use the simpler method for fetching the
root directory path during unit test execution.

Signed-off-by: Joe Stringer <joe@cilium.io>

brb commented on pkg/elf/elf_test.go:
```diff
@@ -0,0 +1,216 @@
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		intOptions := make(map[string]uint32)
```
Wouldn't make sense to stress test with non-empty `intOptions` and `strOptions`, as `elf.copy(...)` returns early if both are empty?

joestringer replied:
Yep, for a more realistic benchmark the PR #7095 adds a full endpoint ELF write benchmark to pkg/data path/loader :)

brb reviewed (approved):
Did the first round of a review - amazing work!
